### ISSUE TITLE
Clippy failed to automatically apply fix: type `bool` cannot be dereferenced

### ISSUE THREAD

**Issue by twirrim** [rust-lang/rust#91561]
Clippy tried to fix some of my code and ran in to issues, and suggested I file a bug report.  I've made a stripped back reproduction of what I saw. 

I tried this code:

```rust
fn main() {
    let input = ["once  upon a time", "there  was an example"];
    let mut input_iter = input.iter();
    let mut words = vec![];
    for _ in 0..input.len() {
        words.push(
            input_iter
                .next()
                .unwrap()
                .split(' ')
                .filter(|x| *x != "")
                .collect::<Vec<_>>(),
        );
    }
    println!("{:?}", words);
}
```
also accessible via playground here:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=829b4f5637a0abf35608926439f1d1a6

Clippy reported:
```
  --> src/main.rs:11:29
   |
11 |                 .filter(|x| *x != "")
   |                             ^^^^^^^^ help: using `!is_empty` is clearer and more explicit: `!*x.is_empty()`
   |
   = note: `#[warn(clippy::comparison_to_empty)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#comparison_to_empty
```

when applying the fix clippy discovered:

```
The following errors were reported:
error[E0614]: type `bool` cannot be dereferenced
  --> src/main.rs:11:30
   |
11 |                 .filter(|x| !*x.is_empty())
   |                              ^^^^^^^^^^^^^

error: aborting due to previous error
```

### Meta

`rustc --version --verbose`:
```
rustc 1.57.0 (f1edd0429 2021-11-29)
binary: rustc
commit-hash: f1edd0429582dd29cccacaf50fd134b05593bd9c
commit-date: 2021-11-29
host: x86_64-unknown-linux-gnu
release: 1.57.0
LLVM version: 13.0.0
```

Also happens in nightly:
```
 rustc --version --verbose
rustc 1.59.0-nightly (efec54529 2021-12-04)
binary: rustc
commit-hash: efec545293b9263be9edfb283a7aa66350b3acbf
commit-date: 2021-12-04
host: x86_64-unknown-linux-gnu
release: 1.59.0-nightly
LLVM version: 13.0.0
```

<details><summary>Backtrace</summary>
<p>

Backtrace doesn't do anything in this case.

```
<backtrace>
```

</p>
</details>

**Comment by armanriazi** [comment]
Try to use  .as_mut() it would be fix it

```

fn main() {
    let input = ["once  upon a time", "there  was an example"];
    let mut input_iter = input.iter();
    let mut words = vec![];
    for _ in 0..input.len() {
        words.push(
            input_iter
                .next()
                .as_mut()
                .unwrap()
                .split(' ')
                .filter(|x| *x != "")
                .collect::<Vec<_>>(),
        );
    }
    println!("{:?}", words);
}
```

**Comment by fmease** [comment]
Clippy now correctly replaces `*x != ""` with `!x.is_empty()`.
In the future, please report Clippy issues over at [rust-lang/rust-clippy](https://github.com/rust-lang/rust-clippy).
